Prepare for Protobuf v22.x

Protobuf v22 has breaking changes affecting the generator (or at least its tests).

```
Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
In file included from ./generator/testing/descriptor_pool_fixture.h:18,
                 from generator/internal/format_method_comments_test.cc:16:
./generator/testing/fake_source_tree.h:34:46: error: 'google::protobuf::io::ZeroCopyInputStream* google::cloud::generator_testing::FakeSourceTree::Open(const std::string&)' marked 'override', but does not override
   34 |   google::protobuf::io::ZeroCopyInputStream* Open(
      |                                              ^~~~
./generator/testing/descriptor_pool_fixture.h:76:18: error: cannot declare field 'google::cloud::generator_testing::DescriptorPoolFixture::source_tree_' to be of abstract type 'google::cloud::generator_testing::FakeSourceTree'
   76 |   FakeSourceTree source_tree_;
      |                  ^~~~~~~~~~~~
./generator/testing/fake_source_tree.h:27:7: note:   because the following virtual functions are pure within 'google::cloud::generator_testing::FakeSourceTree':
   27 | class FakeSourceTree : public google::protobuf::compiler::SourceTree {
      |       ^~~~~~~~~~~~~~
In file included from ./generator/testing/fake_source_tree.h:18:
bazel-out/k8-fastbuild/bin/external/com_google_protobuf/src/google/protobuf/compiler/_virtual_includes/importer/google/protobuf/compiler/importer.h:248:36: note:     'virtual google::protobuf::io::ZeroCopyInputStream* google::protobuf::compiler::SourceTree::Open(absl::lts_20230125::string_view)'
  248 |   virtual io::ZeroCopyInputStream* Open(absl::string_view filename) = 0;
      |                                    ^~~~
```

Note the change from `std::string const&` to `abs::string_view`.

Another one:

```
google/cloud/internal/rest_stub_helpers.cc:50:19: error: 'Status' is not a member of 'google::protobuf::util'
   50 |   protobuf::util::Status proto_to_json_status =
      |                   ^~~~~~
google/cloud/internal/rest_stub_helpers.cc:50:19: note: suggested alternatives:
In file included from bazel-out/k8-fastbuild/bin/external/com_google_protobuf/src/google/protobuf/json/_virtual_includes/json/google/protobuf/json/json.h:38,
                 from bazel-out/k8-fastbuild/bin/external/com_google_protobuf/src/google/protobuf/util/_virtual_includes/json_util/google/protobuf/util/json_util.h:37,
                 from ./google/cloud/internal/rest_stub_helpers.h:25,
                 from google/cloud/internal/rest_stub_helpers.cc:15:
external/com_google_absl/absl/status/status.h:424:7: note:   'absl::lts_20230125::Status'
```

--- a/generator/generator_test.cc
+++ b/generator/generator_test.cc
@@ -15,7 +15,9 @@
 #include "generator/generator.h"
 #include "generator/internal/printer.h"
 #include "generator/testing/error_collectors.h"
+#include "generator/testing/fake_source_tree.h"
 #include "generator/testing/printer_mocks.h"
+#include "google/cloud/internal/type_traits.h"
 #include "google/cloud/log.h"
 #include <google/protobuf/compiler/importer.h>
 #include <google/protobuf/descriptor.h>
@@ -33,24 +35,6 @@ using ::google::protobuf::FileDescriptor;
 using ::google::protobuf::FileDescriptorProto;
 using ::testing::HasSubstr;
 using ::testing::Return;
-
-class StringSourceTree : public google::protobuf::compiler::SourceTree {
- public:
-  explicit StringSourceTree(std::map<std::string, std::string> files)
-      : files_(std::move(files)) {}
-
-  google::protobuf::io::ZeroCopyInputStream* Open(
-      std::string const& filename) override {
-    auto iter = files_.find(filename);
-    return iter == files_.end() ? nullptr
-                                : new google::protobuf::io::ArrayInputStream(
-                                      iter->second.data(),
-                                      static_cast<int>(iter->second.size()));
-  }
-
- private:
-  std::map<std::string, std::string> files_;
-};
 
 char const* const kSuccessServiceProto =
     "syntax = \"proto3\";\n"
@@ -77,7 +61,7 @@ class GeneratorTest : public ::testing::Test {
  private:
   FileDescriptorProto file_proto_;
   generator_testing::ErrorCollector collector_;
-  StringSourceTree source_tree_;
+  generator_testing::FakeSourceTree source_tree_;
   google::protobuf::SimpleDescriptorDatabase simple_db_;
   google::protobuf::compiler::SourceTreeDescriptorDatabase source_tree_db_;
   google::protobuf::MergedDescriptorDatabase merged_db_;

--- a/generator/generator_test.cc
+++ b/generator/generator_test.cc
@@ -17,7 +17,6 @@
 #include "generator/testing/error_collectors.h"
 #include "generator/testing/fake_source_tree.h"
 #include "generator/testing/printer_mocks.h"
-#include "google/cloud/internal/type_traits.h"
 #include "google/cloud/log.h"
 #include <google/protobuf/compiler/importer.h>
 #include <google/protobuf/descriptor.h>

--- a/generator/internal/predicate_utils_test.cc
+++ b/generator/internal/predicate_utils_test.cc
@@ -14,6 +14,7 @@
 
 #include "generator/internal/predicate_utils.h"
 #include "generator/testing/error_collectors.h"
+#include "generator/testing/fake_source_tree.h"
 #include "google/cloud/log.h"
 #include <google/protobuf/compiler/importer.h>
 #include <google/protobuf/descriptor.h>
@@ -720,24 +721,6 @@ TEST(PredicateUtilsTest, PaginationExactlyOneRepatedStringResponse) {
   EXPECT_EQ(result->second, nullptr);
 }
 
-class StringSourceTree : public google::protobuf::compiler::SourceTree {
- public:
-  explicit StringSourceTree(std::map<std::string, std::string> files)
-      : files_(std::move(files)) {}
-
-  google::protobuf::io::ZeroCopyInputStream* Open(
-      std::string const& filename) override {
-    auto iter = files_.find(filename);
-    return iter == files_.end() ? nullptr
-                                : new google::protobuf::io::ArrayInputStream(
-                                      iter->second.data(),
-                                      static_cast<int>(iter->second.size()));
-  }
-
- private:
-  std::map<std::string, std::string> files_;
-};
-
 char const* const kStreamingServiceProto =
     "syntax = \"proto3\";\n"
     "package google.protobuf;\n"
@@ -795,7 +778,7 @@ class StreamingReadTest : public testing::Test {
  private:
   FileDescriptorProto file_proto_;
   generator_testing::ErrorCollector collector_;
-  StringSourceTree source_tree_;
+  generator_testing::FakeSourceTree source_tree_;
   google::protobuf::SimpleDescriptorDatabase simple_db_;
   google::protobuf::compiler::SourceTreeDescriptorDatabase source_tree_db_;
   google::protobuf::MergedDescriptorDatabase merged_db_;

--- a/generator/internal/service_code_generator_test.cc
+++ b/generator/internal/service_code_generator_test.cc
@@ -14,6 +14,7 @@
 
 #include "generator/internal/service_code_generator.h"
 #include "generator/testing/error_collectors.h"
+#include "generator/testing/fake_source_tree.h"
 #include "generator/testing/printer_mocks.h"
 #include "google/cloud/log.h"
 #include <google/protobuf/compiler/importer.h>
@@ -36,24 +37,6 @@ using ::google::protobuf::FileDescriptorProto;
 using ::testing::Contains;
 using ::testing::IsEmpty;
 using ::testing::Return;
-
-class StringSourceTree : public google::protobuf::compiler::SourceTree {
- public:
-  explicit StringSourceTree(std::map<std::string, std::string> files)
-      : files_(std::move(files)) {}
-
-  google::protobuf::io::ZeroCopyInputStream* Open(
-      std::string const& filename) override {
-    auto iter = files_.find(filename);
-    return iter == files_.end() ? nullptr
-                                : new google::protobuf::io::ArrayInputStream(
-                                      iter->second.data(),
-                                      static_cast<int>(iter->second.size()));
-  }
-
- private:
-  std::map<std::string, std::string> files_;
-};
 
 class TestGenerator : public ServiceCodeGenerator {
  public:
@@ -379,7 +362,7 @@ class HasMessageWithMapFieldTest : public testing::Test {
  private:
   FileDescriptorProto file_proto_;
   generator_testing::ErrorCollector collector_;
-  StringSourceTree source_tree_;
+  generator_testing::FakeSourceTree source_tree_;
   google::protobuf::SimpleDescriptorDatabase simple_db_;
   google::protobuf::compiler::SourceTreeDescriptorDatabase source_tree_db_;
   google::protobuf::MergedDescriptorDatabase merged_db_;
@@ -492,7 +475,7 @@ class MethodSignatureWellKnownProtobufTypeIncludesTest : public testing::Test {
  private:
   FileDescriptorProto file_proto_;
   generator_testing::ErrorCollector collector_;
-  StringSourceTree source_tree_;
+  generator_testing::FakeSourceTree source_tree_;
   google::protobuf::SimpleDescriptorDatabase simple_db_;
   google::protobuf::compiler::SourceTreeDescriptorDatabase source_tree_db_;
   google::protobuf::MergedDescriptorDatabase merged_db_;
@@ -584,7 +567,7 @@ class StreamingReadTest : public testing::Test {
  private:
   FileDescriptorProto file_proto_;
   generator_testing::ErrorCollector collector_;
-  StringSourceTree source_tree_;
+  generator_testing::FakeSourceTree source_tree_;
   google::protobuf::SimpleDescriptorDatabase simple_db_;
   google::protobuf::compiler::SourceTreeDescriptorDatabase source_tree_db_;
   google::protobuf::MergedDescriptorDatabase merged_db_;
@@ -660,8 +643,9 @@ service Service1 {
 }
 )""";
 
-  StringSourceTree source_tree(std::map<std::string, std::string>{
-      {"google/cloud/foo/streaming.proto", kBidirStreamingServiceProto}});
+  generator_testing::FakeSourceTree source_tree(
+      std::map<std::string, std::string>{
+          {"google/cloud/foo/streaming.proto", kBidirStreamingServiceProto}});
   google::protobuf::compiler::SourceTreeDescriptorDatabase source_tree_db(
       &source_tree);
   google::protobuf::SimpleDescriptorDatabase simple_db;
@@ -738,8 +722,9 @@ service Service1 {
 }
 )""";
 
-  StringSourceTree source_tree(std::map<std::string, std::string>{
-      {"google/cloud/foo/streaming.proto", kBidirStreamingServiceProto}});
+  generator_testing::FakeSourceTree source_tree(
+      std::map<std::string, std::string>{
+          {"google/cloud/foo/streaming.proto", kBidirStreamingServiceProto}});
   google::protobuf::compiler::SourceTreeDescriptorDatabase source_tree_db(
       &source_tree);
   google::protobuf::SimpleDescriptorDatabase simple_db;
@@ -840,11 +825,12 @@ service Service1 {
 }
 )""";
 
-  StringSourceTree source_tree(std::map<std::string, std::string>{
-      {"google/api/http.proto", kHttpProto},
-      {"google/api/routing.proto", kRoutingProto},
-      {"google/cloud/foo/explicit_routing_service.proto",
-       kExplicitRoutingServiceProto}});
+  generator_testing::FakeSourceTree source_tree(
+      std::map<std::string, std::string>{
+          {"google/api/http.proto", kHttpProto},
+          {"google/api/routing.proto", kRoutingProto},
+          {"google/cloud/foo/explicit_routing_service.proto",
+           kExplicitRoutingServiceProto}});
   google::protobuf::compiler::SourceTreeDescriptorDatabase source_tree_db(
       &source_tree);
   google::protobuf::SimpleDescriptorDatabase simple_db;

--- a/generator/testing/fake_source_tree.cc
+++ b/generator/testing/fake_source_tree.cc
@@ -26,8 +26,8 @@ void FakeSourceTree::Insert(std::string const& filename, std::string contents) {
 }
 
 google::protobuf::io::ZeroCopyInputStream* FakeSourceTree::Open(
-    std::string const& filename) {
-  auto iter = files_.find(filename);
+    FilenameType filename) {
+  auto iter = files_.find(std::string(filename));
   return iter == files_.end()
              ? nullptr
              : new google::protobuf::io::ArrayInputStream(

--- a/generator/testing/fake_source_tree.h
+++ b/generator/testing/fake_source_tree.h
@@ -15,6 +15,8 @@
 #ifndef GOOGLE_CLOUD_CPP_GENERATOR_TESTING_FAKE_SOURCE_TREE_H
 #define GOOGLE_CLOUD_CPP_GENERATOR_TESTING_FAKE_SOURCE_TREE_H
 
+#include "google/cloud/internal/type_traits.h"
+#include "absl/strings/string_view.h"
 #include <google/protobuf/compiler/importer.h>
 #include <google/protobuf/io/zero_copy_stream.h>
 #include <map>
@@ -24,6 +26,18 @@ namespace google {
 namespace cloud {
 namespace generator_testing {
 
+template <typename T, typename AlwaysVoid = void>
+struct SourceTreeOpenArgType {
+  using type = std::string const&;
+};
+
+template <typename T>
+struct SourceTreeOpenArgType<
+    T, google::cloud::internal::void_t<decltype(std::declval<T&>().Open(
+           absl::string_view{}))>> {
+  using type = absl::string_view;
+};
+
 class FakeSourceTree : public google::protobuf::compiler::SourceTree {
  public:
   FakeSourceTree() = default;
@@ -31,8 +45,11 @@ class FakeSourceTree : public google::protobuf::compiler::SourceTree {
 
   void Insert(std::string const& filename, std::string contents);
 
+  using FilenameType = typename generator_testing::SourceTreeOpenArgType<
+      google::protobuf::compiler::SourceTree>::type;
+
   google::protobuf::io::ZeroCopyInputStream* Open(
-      std::string const& filename) override;
+      FilenameType filename) override;
 
  private:
   std::map<std::string, std::string> files_;

--- a/google/cloud/internal/rest_stub_helpers.cc
+++ b/google/cloud/internal/rest_stub_helpers.cc
@@ -27,7 +27,7 @@ Status RestResponseToProto(google::protobuf::Message& destination,
   }
   auto json_response =
       rest_internal::ReadAll(std::move(rest_response).ExtractPayload());
-  if (!json_response.ok()) return json_response.status();
+  if (!json_response.ok()) return std::move(json_response).status();
   google::protobuf::util::JsonParseOptions parse_options;
   parse_options.ignore_unknown_fields = true;
   auto json_to_proto_status = google::protobuf::util::JsonStringToMessage(
@@ -47,11 +47,14 @@ Status RestResponseToProto(google::protobuf::Message& destination,
 
 Status ProtoRequestToJsonPayload(google::protobuf::Message const& request,
                                  std::string& json_payload) {
-  protobuf::util::Status proto_to_json_status =
+  auto proto_to_json_status =
       protobuf::util::MessageToJsonString(request, &json_payload);
   if (!proto_to_json_status.ok()) {
-    return Status{
-        StatusCode::kInternal, std::string{proto_to_json_status.message()}, {}};
+    return internal::InternalError(
+        std::string(proto_to_json_status.message()),
+        GCP_ERROR_INFO()
+            .WithReason("Failure converting proto request to HTTP")
+            .WithMetadata("message_type", request.GetTypeName()));
   }
   return {};
 }

